testplan/testing/multitest/base.py

```diff
@@ -310,432 +275,580 @@ def dry_run(self, status=None):
+        """Callable to be executed after each step."""
+        exceptions = None
+        if step == self.resources.start:
+            exceptions = self.resources.start_exceptions
+        elif step == self.resources.stop:
+            exceptions = self.resources.stop_exceptions
+        if exceptions:
+            for msg in exceptions.values():
+                self.result.report.logger.error(msg)
+            self.result.report.status_override = testplan.report.Status.ERROR
+
+        if step == self.resources.stop:
+            drivers = set(self.resources.start_exceptions.keys())
+            drivers.update(self.resources.stop_exceptions.keys())
+            for driver in drivers:
+                if driver.cfg.report_errors_from_logs:
+                    error_log = os.linesep.join(driver.fetch_error_log())
+                    if error_log:
+                        self.result.report.logger.error(error_log)
+
+    def pre_resource_steps(self):
+        """Runnable steps to be executed before environment starts."""
+        self._add_step(self.make_runpath_dirs)
+        if self.cfg.before_start:
+            self._add_step(
+                self._wrap_run_step(
+                    label="before_start", func=self.cfg.before_start
+                )
+            )
+
+    def main_batch_steps(self):
+        """Runnable steps to be executed while environment is running."""
+        if self.cfg.after_start:
+            self._add_step(
+                self._wrap_run_step(
+                    label="after_start", func=self.cfg.after_start
+                )
+            )
+
+        self._add_step(self.run_tests)
+        self._add_step(self.propagate_tag_indices)
+
+        if self.cfg.before_stop:
+            self._add_step(
+                self._wrap_run_step(
+                    label="before_stop", func=self.cfg.before_stop
+                )
+            )
+
+    def post_resource_steps(self):
+        """Runnable steps to be executed after environment stops."""
+        if self.cfg.after_stop:
+            self._add_step(
+                self._wrap_run_step(
+                    label="after_stop", func=self.cfg.after_stop
+                )
+            )
+
+        self._add_step(self.append_pre_post_step_report)
+
+    def should_run(self):
+        """
+        MultiTest filters are applied in `get_test_context`
+        so we just check if `test_context` is not empty."""
+        return bool(self.test_context)
+
+    def aborting(self):
+        """Suppressing not implemented debug log from parent class."""
+
+    @property
+    def _thread_pool_size(self):
+        """
+        :return: the size of thread pool to use, based on configured limits
+        """
+        if self.cfg.thread_pool_size > 0:
+            return min(
+                self.cfg.thread_pool_size, self.cfg.max_thread_pool_size
+            )
+        else:
+            return max(self.cfg.max_thread_pool_size // 2, 2)
```

Comment:
Seems we could be less conservative here, no? Like if `thread_pool_size <= 0` then give at least 5, since it seems that's the minimum default thread pool size in the [doc](https://docs.python.org/3.7/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor):

> ... If max_workers is None or not given, it will default to the number of processors on the machine, multiplied by 5 ...

Reply:
Oops sorry I missed this comment before merging... I agree, I'll increase the default value.